### PR TITLE
K8SPXC-479 fix update issue

### PR DESCRIPTION
**Pull request by hors**
[K8SPXC-479](https://jira.percona.com/browse/K8SPXC-479)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| storage | passed |
| big-data | passed |
| haproxy | passed |
| scaling | passed |
| self-healing | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| scaling-proxysql | passed |
| limits | passed |
| upgrade | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| upgrade-consistency | failed |
| demand-backup | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| security-context | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scheduled-backup | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| demand-backup-encrypted-with-tls | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| storage | passed |
| scaling | passed |
| recreate | passed |
| big-data | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | failed |
| limits | passed |
| self-healing | passed |
| upgrade-consistency | passed |
| monitoring | passed |
| security-context | failed |
| demand-backup | failed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | failed |
| scheduled-backup | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| demand-backup-encrypted-with-tls | failed |

File: e2e-tests/upgrade-consistency/compare/statefulset_some-name-proxysql-160-oc.yml [hunk 1]
```diff
@@ -150,8 +150,8 @@ spec:
       securityContext:
         supplementalGroups:
           - 1001
-      serviceAccount: default
-      serviceAccountName: default
+      serviceAccount: percona-xtradb-cluster-operator-workload
```
Comment (delgod) [inline e2e-tests/upgrade-consistency/compare/statefulset_some-name-proxysql-160-oc.yml:153]: it should be a default here

File: e2e-tests/upgrade-consistency/compare/statefulset_some-name-proxysql-160.yml [hunk 2]
```diff
@@ -151,8 +151,8 @@ spec:
         fsGroup: 1001
         supplementalGroups:
           - 1001
-      serviceAccount: default
-      serviceAccountName: default
```
Comment (delgod) [inline e2e-tests/upgrade-consistency/compare/statefulset_some-name-proxysql-160.yml:155]: it should be a default here

File: e2e-tests/upgrade-consistency/compare/statefulset_some-name-pxc-160-oc.yml [hunk 3]
```diff
@@ -155,8 +155,8 @@ spec:
       securityContext:
         supplementalGroups:
           - 1001
-      serviceAccount: default
```
Comment (delgod) [inline e2e-tests/upgrade-consistency/compare/statefulset_some-name-pxc-160-oc.yml:158]: it should be a default here

File: e2e-tests/upgrade-consistency/compare/statefulset_some-name-pxc-160.yml [hunk 4]
```diff
@@ -159,8 +159,8 @@ spec:
         fsGroup: 1001
         supplementalGroups:
           - 1001
-      serviceAccount: default
```
Comment (delgod) [inline e2e-tests/upgrade-consistency/compare/statefulset_some-name-pxc-160.yml:162]: it should be a default here

File: pkg/controller/pxc/controller.go [hunk 5]
```diff
@@ -538,11 +544,16 @@ func (r *ReconcilePerconaXtraDBCluster) deploy(cr *api.PerconaXtraDBCluster) err
 
 	inits := []corev1.Container{}
 	if cr.CompareVersionWith("1.5.0") >= 0 {
+		currentVer := "1.5.0"
```
Comment (delgod) [inline pkg/controller/pxc/controller.go:547]:
```suggestion
		imageName := operatorPod.Spec.Containers[0].Image
		if cr.CompareVersionWith(cr.Spec.CRVersion) != 0 {
		    imageName = imageName[imageName,":"),-1] + cr.Spec.CRVersion
		}
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/pxc/controller.go (hunk 5 above)
Comment (hors) [reply pkg/controller/pxc/controller.go]: ok

**Review by delgod — DISMISSED** [review]

File: pkg/controller/pxc/controller.go [hunk 6]
```diff
@@ -265,11 +265,15 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 
 	inits := []corev1.Container{}
 	if o.CompareVersionWith("1.5.0") >= 0 {
+		imageName := operatorPod.Spec.Containers[0].Image
+		if o.CompareVersionWith(o.Spec.CRVersion) != 0 {
```
Comment (qJkee) [inline pkg/controller/pxc/controller.go:269]: comparing current version with current =)

File: pkg/controller/pxc/controller.go [hunk 7]
```diff
@@ -538,11 +542,15 @@ func (r *ReconcilePerconaXtraDBCluster) deploy(cr *api.PerconaXtraDBCluster) err
 
 	inits := []corev1.Container{}
 	if cr.CompareVersionWith("1.5.0") >= 0 {
+		imageName := operatorPod.Spec.Containers[0].Image
+		if cr.CompareVersionWith(cr.Spec.CRVersion) != 0 {
```
Comment (qJkee) [inline pkg/controller/pxc/controller.go:546]: same

File: pkg/pxc/app/statefulset/haproxy.go [hunk 8]
```diff
@@ -106,7 +96,18 @@ func (c *HAProxy) AppContainer(spec *api.PodSpec, secrets string, cr *api.Percon
 		SecurityContext: spec.ContainerSecurityContext,
 	}
 
+	redinessDelay := int32(15)
+	if spec.ReadinessInitialDelaySeconds != nil {
```
Comment (qJkee) [inline pkg/pxc/app/statefulset/haproxy.go:100]: better to move this part to the if below

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

File: pkg/pxc/app/statefulset/haproxy.go (hunk 8 above)
Comment (hors) [reply pkg/pxc/app/statefulset/haproxy.go]: done

**Review by qJkee — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| big-data | passed |
| storage | passed |
| haproxy | passed |
| recreate | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-advanced | passed |
| upgrade-consistency | passed |
| monitoring | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scheduled-backup | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-623-9c021824

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| haproxy | passed |
| recreate | passed |
| scaling | passed |
| storage | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scheduled-backup | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager-ref | passed |

**Pull request merged**